### PR TITLE
Handle Maximum Payment Size

### LINKED ISSUE #3: Track exchange rate and Maximum Payment Size
It should automatically adjust the chunk size based on how much more the other side will accept and the path's MPS.

### PR DESCRIPTION
**Pull request by emschwartz**
adjust the chunk size if it gets F08 errors back

depends on https://github.com/interledgerjs/ilp-protocol-psk2/pull/25
based on https://github.com/interledger/rfcs/issues/397
along with https://github.com/emschwartz/ilp-protocol-paystream/pull/13, resolves https://github.com/emschwartz/ilp-protocol-paystream/issues/3

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -324,21 +327,30 @@ export class PaymentSocket extends EventEmitter {
...
+      await this.maybeSend()
+    } catch (err) {
+      this.debug('error while sending:', err)
+      this.emit('error', err)
+    }
+  }
+
+  protected async maybeSend (): Promise<void> {
+    let shouldContinue = true
+    this.sending = true
```
Comment (sentientwaffle) [inline src/index.ts]: `this.sending` was just set in `startSending()`.

File: src/index.ts [hunk 2]
```diff
@@ -397,8 +409,8 @@ export class PaymentSocket extends EventEmitter {
     } else if (this.peerExpects.isGreaterThan(0)) {
       // Peer is requesting payment, so pay until we've reached our minBalance
       // TODO adjust based on exchange rate and how much peer wants
-      sourceAmount = BigNumber.min(this._balance.minus(this._minBalance), this.maxPaymentSize)
-      sourceAmount = forceValueToBeBetween(sourceAmount, 0, this.maxPaymentSize)
+      sourceAmount = BigNumber.min(this._balance.minus(this._minBalance), this.nextMaxAmount)
+      sourceAmount = forceValueToBeBetween(sourceAmount, 0, this.nextMaxAmount)
 
       // Adjust how much we're sendign to the receiver's minimum
```
Comment (sentientwaffle) [inline src/index.ts]: typo: "sending"

**Review by sentientwaffle — CHANGES_REQUESTED** [review]

File: src/index.ts [hunk 3]
```diff
@@ -495,6 +509,38 @@ export class PaymentSocket extends EventEmitter {
...
+      // Use the F08 error data to determine the path Maximum Payment Size
+      const reader = oer.Reader.from(result.unauthenticatedData)
+      try {
+        const receivedAmount = highLowToBigNumber(reader.readUInt64())
+        const maximumAmount = highLowToBigNumber(reader.readUInt64())
+
+        // If we got here without an error being thrown that means the connector
+        // did actually send those values back (thanks!)
+        const scalingFactor = maximumAmount.dividedBy(receivedAmount)
+        this.maxPaymentSize = sourceAmount.times(scalingFactor).decimalPlaces(0, BigNumber.ROUND_FLOOR)
```
Comment (sharafian) [inline src/index.ts:539]: what would happen if this amount came out to zero?

**Review by sentientwaffle — APPROVED** [review]

File: src/index.ts (hunk 3 above)
Comment (emschwartz) [reply src/index.ts:539]: Bad things. It sends forever. Fixed in https://github.com/emschwartz/ilp-protocol-paystream/pull/17/commits/e52ad07262518600a645283e00b876d50ff32d81

**Review by sharafian — APPROVED** [review]

**Pull request merged**